src/driver/express/ExpressDriver.ts

```diff
@@ -117,15 +118,19 @@ export class ExpressDriver extends BaseDriver implements Driver {
                 if (isPromiseLike(checkResult)) {
                     checkResult.then(result => {
                         if (!result) {
-                            return this.handleError(new AccessDeniedError(action), actionMetadata, action);
+                            return this.handleError((actionMetadata.authorizedRoles.length === 0
+                                    ? new AuthorizationRequiredError(action) : new AccessDeniedError(action)),
+                                actionMetadata, action);
 
                         } else {
                             next();
                         }
                     });
                 } else {
                     if (!checkResult) {
-                        this.handleError(new AccessDeniedError(action), actionMetadata, action);
+                        this.handleError((actionMetadata.authorizedRoles.length === 0
```

Comment:
The only traumatic thing is the hypocrisy:
- `Promise.resolve` removes half on duplicated code block but it's slower, so we stick with messy code because "its important to make it performant as we can"
- now "all these code blocks become messy", so we should remove ternary and tail call optimization to "beatify code"

So please be consistent - peformance at any cost or readable and maintainable code 😄